### ISSUE TITLE
What does the second argument in: "auto mynet = make_mlp<tan_h>({ 32 * 32, 300, 10 });" (number 300)? The number of layers or neurons?

### ISSUE THREAD

**Issue by dimanys111** [tiny-dnn/tiny-dnn#575]
What does the second argument in: "auto mynet = make_mlp<tan_h>({ 32 * 32, 300, 10 });" (number
 300)?
The number of layers or neurons?

**Comment by Randl** [comment]
Please, don't spam with issues ( https://github.com/tiny-dnn/tiny-dnn/issues/571 ) . If you have follow up questions, you can ask them there even if the issue is closed. 
Answering you question, yes, it is number of neurons in a middle layer. Check https://github.com/tiny-dnn/tiny-dnn/blob/master/docs/how_tos/How-Tos.md and other docs for more information

**Comment by dimanys111** [comment]
auto mynet = make_mlp<tan_h>({ 32 * 32, 100, 100, 10 });
It is two layers of neurons at 100?

**Comment by Randl** [comment]
Yes, you can check the implementation https://github.com/tiny-dnn/tiny-dnn/blob/master/tiny_dnn/io/layer_factory.h

**Comment by dimanys111** [comment]
And from each input neuron synapses are to each neuron of the second layer?
32 * 32 * 100 = 102400 synapses with the coefficient? (In my example)

**Comment by Randl** [comment]
Yes

**Comment by dimanys111** [comment]
mynet.train<mse>(optimizer, train_images, train_labels, n_minibatch, n_train_epochs, on_enumerate_minibatch, on_enumerate_epoch);
And what role is played by "n_minibach"? It should be chosen for a better training?
I watched the source code, but has not yet figured out.

**Comment by Randl** [comment]
It is size of minibatch, as you can see from the name. See, for example https://www.quora.com/What-is-the-difference-between-batch-mode-and-mini-batch-in-machine-learning
It is number of sample proceeded simultaneously.

**Comment by dimanys111** [comment]
And treated "n_minibah" elements in a separate thread?

**Comment by dimanys111** [comment]
Why in a class of hyperbolic tangent derivative = float_t (1) - sqr (y)?
In various sources it is 
[http://tehtab.ru/netcat_files/Image/GuideMathematics/DifferentialsAndIntegrals/DifferentialsTable/Differentials_clip_image027.gif](url)

`class tan_h : public function {
 public:
  using function::df;

  inline float_t f(const vec_t &v, size_t i) const override {
    return std::tanh(v[i]);
  }

  inline void itef(vec_t &out, const vec_t &in, size_t cnt) const {
    for (size_t i = 0; i < cnt; i++) {
      out[i] = std::tanh(in[i]);
    }
  }

  // fast approximation of tanh (improve 2-3% speed in LeNet-5)
  /*float_t f(float_t x) const {
      const float_t x2 = x * x;
      x *= 1.0 + x2 * (0.1653 + x2 * 0.0097);
      return x / std::sqrt(1.0 + x * x);// invsqrt(static_cast<float>(1.0 + x
  *
  x));
  }*/

  inline float_t df(float_t y) const override { return float_t(1) - sqr(y); }
  void itedf(vec_t &cur,
             const vec_t &prev,
             const vec_t &out,
             size_t cnt) const {
    for (size_t i = 0; i < cnt; i++) {
      cur[i] = prev[i] * df(out[i]);
    }
  }

  std::pair<float_t, float_t> scale() const override {
    return std::make_pair(float_t(-0.8), float_t(0.8));
  }

 private:
  /*float invsqrt(float x) const {
      float x2 = x * 0.5f;
      long i = *reinterpret_cast<long*>(&x);

      i = 0x5f3759df - (i >> 1);
      x = *reinterpret_cast<float*>(&i);
      x = x * (1.5f - (x2 * x * x));
      return x;
  }*/
};`

**Comment by beru** [comment]
@dimanys111 Please reformat your comment with Syntax highlighting markdown.
Enclose code with 3 backquotes.
```cpp
printf("Hello, World\n");
```
